### PR TITLE
Pull the trace struct out of the low-level interface.

### PR DESCRIPTION
**Pull request by vext01**
This is a companion change to https://github.com/softdevteam/yktrace/pull/1

### CODE REVIEW

File: src/test/run-pass/yk_swt/collect-traces-threaded.rs [hunk 1]
```diff
@@ -25,29 +25,32 @@ fn main() {
...
 
 // Copies a trace into a plain Rust Vec of tuples so we can compare them.
-fn trace_to_vec(t: &SWTrace) -> Vec<(u64, u32, u32)> {
+fn trace_to_vec(tup: (*mut SWTMirLoc, usize)) -> Vec<(u64, u32, u32)> {
+    let (buf, len) = tup;
     let mut v = Vec::new();
-    for i in 0..t.len() {
-        let loc = t.loc(i);
+    for i in 0..len {
+        if i > isize::max_value() as usize {
```
Comment (ltratt) [inline src/test/run-pass/yk_swt/collect-traces-threaded.rs]: Wouldn't it be easier to hoist this outside the loop? Something like `assert!(len < (isize::max_value() as usize))`.

Comment (vext01) [reply src/test/run-pass/yk_swt/collect-traces-threaded.rs]: Indeed!

**Comment by ltratt** [comment]
Which PR should be done first?

**Comment by vext01** [comment]
> Which PR should be done first?

I think we should decide if we like the API outlined in https://github.com/softdevteam/yktrace/pull/1, and if we do, then we merge this ykrustc PR first, then the yktrace PR. This is because the yktrace PR depends on the compiler changes.

BTW, you may have noticed the FIXME. I'd like to restructure our libcore code at some point, but I didn't want to mix it in with this PR and make it confusing to review. I hope that's OK.

**Comment by vext01** [comment]
Your comment was addressed in my recent push BTW, so this is also ready for re-review.

**Comment by ltratt** [comment]
The updates look good, but I guess we'll need to wait for the `Option` change to `start_tracing` to be fixed here too?

**Comment by vext01** [comment]
Correct

**Comment by ltratt** [comment]
Do we need to do something to this PR before considering it for merging?

**Comment by vext01** [comment]
If you have no further comments, we can squash this.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
squashed and revised commit message.

**Comment by vext01** [comment]
```
+ ./x.py clean
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
```

**Comment by vext01** [comment]
bors retry

**Comment by vext01** [comment]
OK it's building on worker1 (it failed on worker4). I restarted all of the workers just in case that's needed to pick up the "fresh clone" settings we changed yesterday.

I'll be keeping an eye on this, as I'm not totally sure if it's fixed it for all workers or not...

**Pull request merged**